tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1591 @@
+/*
+Package WAL implements a write ahead log optimized for write throughput
+that can be put in front of the database index.
+
+The WAL is broken into different partitions. The default number of
+partitions is 5. Each partition consists of a number of segment files.
+By default these files will get up to 2MB in size before a new segment
+file is opened. The files are numbered and start at 1. The number
+indicates the order in which the files should be read on startup to
+ensure data is recovered in the same order it was written.
+
+Partitions are flushed and compacted individually. One of the goals with
+having multiple partitions was to be able to flush only a portion of the
+WAL at a time.
+
+The WAL does not flush everything in a partition when it comes time. It will
```

Comment:
The reason full flushes exist is because once a shard is cold for writes, you want it to get flushed entirely to the BoltDB. Performance isn't really a concern at that point because it's no longer getting written to.

If it's still hot for writes and you hit this threshold, then you have so little data coming in, that performance isn't a concern for you.